brave/brave-ios
# Fix #7887: Sync UX improvement - Conditional Device Number

## Summary of Changes
This pull request aims to add device limitations for a device joining an existing key chain.

The device number limitations can be listed as 

`1..4 devices `
- Join the chain silently and enable bookmarks

`5..9 devices`
 - Show alert with device list; if user answers no - leave chain, if yes - enable the bookmarks type
 
`10 and more` 
- Add alert to block and prevent sync (leave chain)


This pull request fixes #7887

## Submitter Checklist:



## Updated Text:
Text PR _ https://github.com/brave/brave-ios/pull/7984

Text PR2 _ https://github.com/brave/brave-ios/pull/7999

## Test Plan:

Test 1: Create NEW CHAIN
- Create a brand new sync chain
- Check Bookmark is enabled by default
- Add another device to sync chain
- Check Bookmark Sync is working

Test 2: JOIN 1-4 DEVICES
- Create a brand new sync chain
- Add another device to sync chain until reached 4th device
- Check warning- alert isnt presented
- And sync is working properly

Test 3: JOIN 5-9 DEVICES
- Continue adding device to previous chain where 4 devices is added
- Check the 5h device will show an alert indicating the names of devices in sync chain
- Test the scenario for YES for seeing new device is joined and sync bookmark is enabled
- Test the scenario for NO for seeing new device couldnt join the chain
- Continue adding devices this way until reaching 10th

Test 4: JOIN 10 DEVICE
- Add 10 th device to the same chain
- Check error is displayed and the designated device cant be included to this chain

## Screenshots:

Test 3: JOIN 5-9 DEVICES

![121212](https://github.com/brave/brave-ios/assets/6643505/9cbee1c1-b8d4-4f33-af6e-d869f715468a)

Test 3 YES 

https://github.com/brave/brave-ios/assets/6643505/f8e1c891-a8fa-4f61-a620-d3cf2ad3c573

Test 3 NO

https://github.com/brave/brave-ios/assets/6643505/41fc8033-6333-4da9-9502-71725cd601c4

Test 4: JOIN 10 DEVICE

![14122331223332323](https://github.com/brave/brave-ios/assets/6643505/c14c748d-e042-45ab-8a69-a4c548f3404e)

https://github.com/brave/brave-ios/assets/6643505/895c6c72-d2b9-4163-a956-663a9b911a31


## Reviewer Checklist:

  - `QA/(Yes|No)`
  - `bug` / `enhancement`
